### PR TITLE
[Merged by Bors] - refactor: single-edge graph

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
From https://github.com/leanprover-community/mathlib4/pull/9267#discussion_r1451605191:
> I would prefer we use lattice operations for adding edges. The idea is to make constructor `SimpleGraph.edge (v w : V) : SimpleGraph V` that creates a graph with a single edge between `v` and `w` if they're not equal (and no edge if they are), and then `G.addEdge v w` would instead be `G ⊔ edge v w`. This is more versatile, though perhaps lemmas such as `addEdge_of_adj` are a bit more brittle to apply.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Operations.lean [hunk 1]
```diff
@@ -124,31 +125,42 @@ end ReplaceVertex
...
+lemma sup_edge_self : G ⊔ edge s s = G := by
+  rw [edge_self_eq_bot, sup_of_le_left bot_le]
 
 variable {s t}
 
-lemma addEdge_of_adj (h : G.Adj s t) : G.addEdge s t = G := by
-  ext
-  simp only [addEdge, ne_eq, G.ne_of_adj h, not_false_eq_true, true_and, or_iff_left_iff_imp]
-  rintro (_ | _) <;> simp_all [adj_comm]
+lemma edge_edgeSet_of_ne (h : s ≠ t): (edge s t).edgeSet = {s(s, t)} := by
```
Comment (lambda-fairy) [inline Mathlib/Combinatorics/SimpleGraph/Operations.lean:145]: Nit:

```suggestion
lemma edge_edgeSet_of_ne (h : s ≠ t) : (edge s t).edgeSet = {s(s, t)} := by
```

File: Mathlib/Combinatorics/SimpleGraph/Operations.lean [hunk 2]
```diff
@@ -124,31 +125,42 @@ end ReplaceVertex
...
+/-- The graph with a single `s-t` edge. It is empty iff `s = t`. -/
+def edge : SimpleGraph V := fromEdgeSet {s(s, t)}
+
+lemma edge_adj (v w : V) : (edge s t).Adj v w ↔ (v = s ∧ w = t ∨ v = t ∧ w = s) ∧ v ≠ w := by
+  rw [edge, fromEdgeSet_adj, Set.mem_singleton_iff, Sym2.eq_iff]
+
+lemma edge_self_eq_bot : edge s s = ⊥ := by
+  ext a b
+  rw [edge_adj, or_self, bot_adj, iff_false, not_and, and_imp, not_ne_iff]
+  exact fun j k ↦ j.trans k.symm
```
Comment (lambda-fairy) [inline Mathlib/Combinatorics/SimpleGraph/Operations.lean:137]: I see a lot of logical manipulation. Would something like

```lean
  rw [edge_adj, bot_adj]
  tauto
```

work here?

**Review by lambda-fairy — COMMENTED** [review]
This is pretty cool. I assume we can delete an edge with `G \ edge v w` too?

**Review by kmill — APPROVED** [review]
Thanks for following up on this idea. Once you get that spacing fixed, and decide how to address `edge_self_eq_bot`, we'll merge.

**Comment by Parcly-Taxel** [comment]
@kmill done.

File: Mathlib/Combinatorics/SimpleGraph/Operations.lean (hunk 2 above)
Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Operations.lean]: `tauto` doesn't work but `aesop` does.

**Comment by kmill** [comment]
Thanks!

bors r+

**Pull request closed without merging**